File: app/src/main/java/com/battlelancer/seriesguide/lists/ListsDistillationSettings.java
// SPDX-License-Identifier: Apache-2.0
// Copyright 2015-2025 Uwe Trottmann

package com.battlelancer.seriesguide.lists;

import android.content.Context;
import androidx.preference.PreferenceManager;
import com.battlelancer.seriesguide.provider.SeriesGuideContract.ListItems;
import com.battlelancer.seriesguide.settings.DisplaySettings;
import com.battlelancer.seriesguide.shows.ShowsDistillationSettings;
import com.battlelancer.seriesguide.shows.ShowsDistillationSettings.ShowSortOrder;

/**
 * Provides settings used to sort displayed list items.
 */
public class ListsDistillationSettings {

    public static class ListsSortOrderChangedEvent {
    }

    public static String KEY_SORT_ORDER = "com.battlelancer.seriesguide.lists.sortorder";

    /**
     * Builds an appropriate SQL sort statement for sorting lists.
     */
    static String getSortQuery(Context context) {
        int sortOrderId = getSortOrderId(context);

        // convert to show sort order id
        switch (sortOrderId) {
            case ListsSortOrder.LATEST_EPISODE_ID:
                sortOrderId = ShowSortOrder.LATEST_EPISODE_ID;
                break;
            case ListsSortOrder.OLDEST_EPISODE_ID:
                sortOrderId = ShowSortOrder.OLDEST_EPISODE_ID;
                break;
            case ListsSortOrder.LAST_WATCHED_ID:
                sortOrderId = ShowSortOrder.LAST_WATCHED_ID;
                break;
            case ListsSortOrder.LEAST_REMAINING_EPISODES_ID:
                sortOrderId = ShowSortOrder.LEAST_REMAINING_EPISODES_ID;
                break;
            default:
                sortOrderId = ShowSortOrder.TITLE_ID;
                break;
        }

        String baseQuery = ShowsDistillationSettings.getSortQuery2(sortOrderId, false,
                DisplaySettings.isSortOrderIgnoringArticles(context));

        // append sorting by list type
        return baseQuery + "," + ListItems.SORT_TYPE;
    }

    /**
     * Returns the id as of {@link ListsDistillationSettings.ListsSortOrder}
     * of the current list sort order.
     */
    public static int getSortOrderId(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context)
                .getInt(KEY_SORT_ORDER, ListsSortOrder.TITLE_ALPHABETICAL_ID);
    }

    public interface ListsSortOrder {
        int TITLE_ALPHABETICAL_ID = 0;
        // @deprecated Only supporting alphabetical sort order going forward.
        // int TITLE_REVERSE_ALHPABETICAL_ID = 1;
        int LATEST_EPISODE_ID = 2;
        int OLDEST_EPISODE_ID = 3;
        int LAST_WATCHED_ID = 4;
        int LEAST_REMAINING_EPISODES_ID = 5;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/lists/ListsDistillationSettings.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// SPDX-FileCopyrightText: Copyright © 2015 Uwe Trottmann <uwe@uwetrottmann.com>

package com.battlelancer.seriesguide.lists

import android.content.Context
import androidx.core.content.edit
import androidx.preference.PreferenceManager
import com.battlelancer.seriesguide.lists.database.SgListItemWithDetails
import com.battlelancer.seriesguide.settings.DisplaySettings

/**
 * Provides settings used to sort displayed list items.
 */
object ListsDistillationSettings {

    class ListsSortOrderChangedEvent

    private const val KEY_SORT_ORDER = "com.battlelancer.seriesguide.lists.sortorder"

    /**
     * Builds an appropriate SQL sort statement for sorting [SgListItemWithDetails] rows.
     */
    fun getSortQuery(context: Context): String {
        val query = StringBuilder()

        val sortOrder = getSortOrder(context)
        val orderClause: String = when (sortOrder) {
            ListsSortOrder.TITLE_ALPHABETICAL ->
                if (DisplaySettings.isSortOrderIgnoringArticles(context)) {
                    SgListItemWithDetails.SORT_TITLE_NO_ARTICLE
                } else {
                    SgListItemWithDetails.SORT_TITLE
                }

            ListsSortOrder.LATEST_EPISODE -> SgListItemWithDetails.SORT_LATEST_RELEASE_DATE
            ListsSortOrder.OLDEST_EPISODE -> SgListItemWithDetails.SORT_OLDEST_RELEASE_DATE
            ListsSortOrder.LAST_WATCHED -> SgListItemWithDetails.SORT_LAST_WATCHED
            ListsSortOrder.LEAST_REMAINING_EPISODES -> SgListItemWithDetails.SORT_REMAINING_EPISODES
        }
        query.append(orderClause)

        // Then by type
        query.append(",").append(SgListItemWithDetails.SORT_TYPE)

        return query.toString()
    }

    fun saveSortOrder(context: Context, sortOrder: ListsSortOrder) {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putInt(KEY_SORT_ORDER, sortOrder.id)
        }
    }

    /**
     * Returns the current [ListsSortOrder], defaulting to [ListsSortOrder.TITLE_ALPHABETICAL]
     * for unknown stored values.
     */
    fun getSortOrder(context: Context): ListsSortOrder {
        return PreferenceManager.getDefaultSharedPreferences(context)
            .getInt(KEY_SORT_ORDER, ListsSortOrder.TITLE_ALPHABETICAL.id)
            .let { ListsSortOrder.fromId(it) }
    }

    enum class ListsSortOrder(val id: Int) {
        TITLE_ALPHABETICAL(0),

        // Deprecated: Only supporting alphabetical sort order going forward.
        // TITLE_REVERSE_ALPHABETICAL(1),

        LATEST_EPISODE(2),
        OLDEST_EPISODE(3),
        LAST_WATCHED(4),
        LEAST_REMAINING_EPISODES(5);

        companion object {
            fun fromId(id: Int): ListsSortOrder =
                entries.find { it.id == id } ?: TITLE_ALPHABETICAL
        }
    }
}
